mpi.h

```diff
@@ -1196,6 +1181,25 @@ MPI_Session MPI_Session_f2c(MPI_Fint session);
 MPI_Fint MPI_Win_c2f(MPI_Win win);
 MPI_Win MPI_Win_f2c(MPI_Fint win);
 
+/* MPI deprecated types and constants */
+#define MPI_Copy_function              MPI_Comm_copy_attr_function
+#define MPI_Delete_function            MPI_Comm_delete_attr_function
+#define MPI_NULL_COPY_FN               MPI_COMM_NULL_COPY_FN
+#define MPI_DUP_FN                     MPI_COMM_DUP_FN
+#define MPI_NULL_DELETE_FN             MPI_COMM_NULL_DELETE_FN
+
+/* MPI deprecated functions */
+#define MPI_Attr_delete                MPI_Comm_delete_attr
+#define MPI_Attr_get                   MPI_Comm_get_attr
+#define MPI_Attr_put                   MPI_Comm_set_attr
+#define MPI_Get_elements_x             MPI_Get_elements_c
+#define MPI_Keyval_create              MPI_Comm_create_keyval
+#define MPI_Keyval_free                MPI_Comm_free_keyval
+#define MPI_Status_set_elements_x      MPI_Status_set_elements_c
+#define MPI_Type_get_extent_x          MPI_Type_get_extent_c
+#define MPI_Type_get_true_extent_x     MPI_Type_get_true_extent_c
+#define MPI_Type_size_x                MPI_Type_size_c
```

Comment:
I am against this. Either you want to include them, then declare them explicitly and treat them as regular abi symbols; or remove them completely and do not support legacy code.

Reply:
* These symbols cannot be removed in 4.2, they have been just deprecated so far. Moreover, a full removal will do nothing more than harm users.
* But this stuff is deprecated, and we would like to somehow start to get rid of it or discourage use, right? If we keep this stuff, we will probably have to keep it FOREVER. In the future, we are not going to break the ABI compatibility of the library just for removing silly deprecated routines added 30 years ago or workarounds no longer needed like the `_x` routines. As I said in the meeting, MPI 3.0 was the only version to actually remove things and break APIs and ABIs. MPICH did not went along with it. Open MPI did, and that unravel some minor level of misery on end users. I do remember PETSc folks rushing for a workaround to fix the creation of a struct datatype.

This change is a very good compromise between removal, and keeping support forever.
* To use the ABI, applications have to be recompiled anyway. 
* Upon recompilation, most code will not notice in practice that a bunch of long deprecated stuff are now preprocessor aliases. 
* User binaries will reference the new replacement symbols. 
* MPI implementations will not have to care about implementing and supporting this deprecated stuff, at least  under the ABI (not every implementation generates bindings with scripts).

Reply:
You are effectively removing deprecated functions. The aliases are just one of the workaround for patching code. I don't have an opinion one way or another.  This is should be up to the forum to decide.

Reply:
> You are effectively removing deprecated functions.

Yes, I'm removing them from the C ABI, but not quite from the C API, in the sense that user code compiles unmodified to avoid any pain and grief. The produced binaries will not have any references to the old stuff.

> This should be up to the forum to decide.

Definitely. I'm just submitting PRs to open up the discussion, and to make the point with actual code that user applications (mpi4py so far, and only for the `_x` functions) can happily compile.